src/box/relay.cc

```diff
@@ -166,11 +166,11 @@ struct relay {
 	 * separated helps to synchronize the data coming from TX and WAL. Such
 	 * as term bumps from TX with PROMOTE rows from WAL.
 	 */
-	struct cbus_endpoint tx_endpoint;
+	struct cbus_endpoint tx_endpoint[2];
 	/** A pipe from 'relay' thread to 'tx' */
-	struct cpipe tx_pipe;
+	struct cpipe tx_pipe[2];
 	/** A pipe from 'tx' thread to 'relay' */
-	struct cpipe relay_pipe;
+	struct cpipe relay_pipe[2];
```

Comment:
Why do you need a second endpoint for that? I can't see the underlying problem why isn't the first endpoint being handled properly. I see that `relay_yield_and_send_heartbeat()` should be called, so what is the bug? Is the problem that we do not get notifications from the TX thread because we do not call `cbus_process()`? It doesn't look like you need a new endpoint then.

For example, we could create a separate fiber which runs `cbus_loop(tx_endpoint)` and remove `cbus_process()` from all other places. Would it work?

Adding a new endpoint will only make the code more complex due to more things to take care of and will eventually introduce races when we push different stuff into different endpoints and won't have any delivery order guarantees.

Reply:
The first solution that came to mind.

> Adding a new endpoint will only make the code more complex due to more things to take care of and will eventually introduce races ...

Yes, that's true.

Reply:
Done